### ISSUE TITLE
webgpu: batchNorm* related cases fail due to 'The number of storage buffers exceeds the maximum per-stage limit'

### ISSUE THREAD

**Issue by qjia7** [tensorflow/tfjs#3648]
The number of storage buffers exceeds the maximum per-stage limit.
    at ValidateBindingCounts (..<URL>)
    at CreateDefault (..<URL>)
    at CreateComputePipelineInternal (..<URL>)
14Object is an error.
    at ValidateObject (..<URL>)
    at ValidateGetBindGroupLayout (..<URL>)
17Object is an error.
    at ValidateObject (..<URL>)
    at ValidateBindGroupDescriptor (..<URL>)
    at CreateBindGroupInternal (..<URL>)
17Object is an error.
    at Finish (..<URL>)
17Object is an error.
    at ValidateObject (..<URL>)
    at ValidateSubmit (..<URL>)
debug.js:15 FAILED batchNorm4D test-webgpu {"WEBGPU_CPU_FORWARD":false} simple batchnorm4D, no offset, 2x1x1x2

Above failure is due to the change https://dawn-review.googlesource.com/c/dawn/+/24481/6/src/common/Constants.h#42. In those shaders, we used more than 4 storage buffers. It seems that we should use uniform buffer or constant value instead of storage buffer to reduce the binding storage buffer count.

**Comment by qjia7** [comment]
@annxingyuan @kainino0x 
From https://github.com/tensorflow/tfjs/blob/master/tfjs-backend-webgpu/src/kernels/FusedBatchNorm.ts#L43, we can see that FusedBatchNorm has 5 inputs tensors [x as Tensor, mean as Tensor, variance as Tensor, offset as Tensor, scale as Tensor]. If all of them use storage buffer, it will exceed the limitation of webgpu which's max value is 4. However, if we use uniform buffer, it must follow the std140 layout which needs to padding the value in tensor. If the tensor is too big, constant value is also not a good option. Maybe we should consider to use texture to store the last tensor if no storage buffer can be used?

**Comment by Jiawei-Shao** [comment]
Hi @Kangz, in the latest Dawn there is a newly-added limitation on the maximum number of storage buffers inside one shader stage (4), while in the current version of TFJS we actually need 5 storage buffers. Could you give us some idea about this issue?

**Comment by kainino0x** [comment]
This limit is documented here:
https://github.com/gpuweb/gpuweb/blob/main/design/Limits.md
and (like the rest of them) comes from Vulkan.

A limit of 4 is not uncommon on Qualcomm and ARM:
http://vulkan.gpuinfo.org/displaydevicelimit.php?name=maxPerStageDescriptorStorageBuffers&platform=android

**Comment by kainino0x** [comment]
If there's any of these inputs that is always small, and you can convert it to a uniform buffer (or even a texture), it would bring you under the limit.

Another possible approach would be to use copies to merge some of the inputs into a single input storage buffer that you address at offsets.

If this op has 5 inputs, do you need 6 storage buffers (5, plus one for the output)?

**Comment by qjia7** [comment]
>If there's any of these inputs that is always small, and you can convert it to a uniform buffer (or even a texture), it would bring you under the limit.

I don't think we can make sure that those inputs are always small. Per https://www.tensorflow.org/api_docs/cc/class/tensorflow/ops/fused-batch-norm, scale/offset/mean/variance are 1 D tensor and the size of 1D Tensors matches the dimension C of the 4D Tensors. But dimension C can be a large number, for example, 1024 or larger. @annxingyuan Please collect me if I am wrong.
In fact, not only batchNorm* related cases have such kind of issue. Some other cases also have the similar issue. For example, fused-conv2d.
```
    layout(std430, set = 0, binding = 0) writeonly buffer ssbOut {
      float result[];
    };

      layout(std430, set = 0, binding = 1) readonly buffer ssbx {
        float x[];
      };

      layout(std430, set = 0, binding = 2) readonly buffer ssbW {
        float W[];
      };

      layout(std430, set = 0, binding = 3) readonly buffer ssbbias {
        float bias[];
      };

      layout(std430, set = 0, binding = 4) readonly buffer ssbpreluActivationWeights {
        float preluActivationWeights[];
      };
```

>Another possible approach would be to use copies to merge some of the inputs into a single input storage buffer that you address at offsets.

Well, it looks like a doable method. We can merge some parameters weights into one buffer, for example, the bias, the preluActivation. Or we use readonly texture for each of those parameters. I will take a look.

>If this op has 5 inputs, do you need 6 storage buffers (5, plus one for the output)?

Oh, yes. It should need 6 storage buffers if op has 5 inputs.

**Comment by Kangz** [comment]
One thing we could do is add support for `GPULimits` in Chromium so TF.js can ask for increased limits (but not work on devices with the bare minimum). And in the meantime we can set the minimum to 6 in Dawn so that TF.js isn't block on it. WDYT? @austinEng @kainino0x ?

**Comment by austinEng** [comment]
Temporary limit change sounds okay with me since we're not testing or shipping on mobile platforms yet. We'll start failing CTS tests that test the limits though.

**Comment by kainino0x** [comment]
sgtm

**Comment by austinEng** [comment]
fyi: https://dawn-review.googlesource.com/c/dawn/+/25840 should be in Chrome Canary in the next couple of days

**Comment by gaikwadrahul8** [comment]
Hi, @qjia7 

Apologize for the delayed response and we're re-visiting our older issues and checking whether those issues got resolved or not as of now so May I know are you still looking for the solution or your issue got resolved ?

If issue still persists after trying with latest version of TFJs please let us know with error log and code snippet to replicate the same issue from our end ?

Could you please confirm if this issue is resolved for you ? Please feel free to close the issue if it is resolved ? Thank you!

**Comment by qjia7** [comment]
It's not an issue now since `maxStorageBuffersPerShaderStage` is 8.
